Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+* If not specified, the "muxer" (`dotnet.exe`) behavior is used to locate the `hostfxr`
+* If specified, the `apphost` behavior is used
+
+
+## Improve API to run application
+New API will be added to `hostfxr`.
+
+``` C++
+struct hostfxr_runtime_properties
+{
+    int count;
+    const char ** keys;
+    const char ** values;
+};
+
+struct hostfxr_parameters
+{
+    int size;
+    const char * host_path;
+    const char * dotnet_root;
+    const char * app_path;
+    hostfxr_runtime_properties additional_properties;
+};
+
+int hostfxr_main_with_parameters(
+        const int argc,
+        const char* argv[],
+        const hostfxr_parameters * parameters);
+```
+
+This new API supercedes the existing `hostfxr_main` and `hostfxr_main_startupinfo`. The function is the main entry point for running applications and dotnet commands. It encapsulates both the `muxer` functionality for running apps via `dotnet app.dll` or `dotnet exec app.dll`, as well as running SDK commands like `dotnet new` and so on.
+
+The proposed behavior is the same as for the existing entry points `hostfxr_main` and `hostfxr_main_startupinfo` with the addition of `parameters` structure. This is meant to:
+* Add the ability to specify additional runtime properties which will be added to the property bag pass to the runtime during its initialization.
+* Make the API extensible without adding new entry point. In the future the parameters structure can be extended with additional members.
+
+The ability to specify additional runtime properties introduces potential collisions with runtime properties initialized by the hosting layer. This ability is not new, it's already possible to specify arbitrary set of additional runtime properties through the `configProperties` section of the `.runtimeconfig.json`. The existing behavior for collisions is somewhat undefined by it effectively means:
+* Properties supplied by the host are added first
+* Any additional properties are added after these - duplicates are retained
+* Runtime initialization will use the last value of any given property it finds for properties it needs to start running managed code
+* Later on during runtime initialization duplicates will cause a failure and the runtime will fail to start with `0x80070057 (E_INVALIDARG)` (internally the failure is due to adding the same key twice to a `Dictionary<K,V>`).
+
+This means that having duplicate properties leads to runtime initialization failure (although the exact set of properties might end up with different failure modes, but it will always fail).
+
+For now, this proposal is not trying to improve this behavior and the additional properties specified through the new `hostfxr` API will behave exactly like those specified via `.runtimeconfig.json`. Since duplicates are not allowed (they cause failures) there's no need to specify precedence (although the logical one would be to take the properties from the API over those in `.runtimeconfig.json`).
```

Comment:
We either need to support layering or give the host a way to tell if the property was set in runtime config to be able to avoid runtime failure.

Reply:
While I understand the theoretical concern around this, I'm still to find a scenario where it would actually matter.

Another problem is that there's no single "resolution policy" we could apply to all properties. Basically each property would need its own reoslution logic. Given that the host only knows about cca 10 properties, it could (in theory) implemente some resolution logic for those 10, but not the rest.

We could expose some kind of callback to perform conflict resolution I guess, but I'm not sure if that would be useful (no idea who would use it and how). There's also in theory multiple `.runtimeconfig.json` files involved all with possible redefinition of the same property - which the callback would somehow have to handle - which means it would need to be given enough information to be able to do that.

Do you have a scenario in mind which would run into this problem? (As noted above, the problem already exists, the new API just makes it a little bit more probable).

Reply:
Consider https://github.com/dotnet/coreclr/issues/21450, when the property gets added we would want to always set it in AspNetCoreModule to 1M to match dotnet.exe behaviour. But if the customer decides to set this value manually in csproj he would get a runtime error.

Reply:
Would be cool if `hostfxr_initialize` could calculate properties that would've been set by the hostfxr and allow external code to inspect/change them before calling `hostfxr_main_with_parameters`

Reply:
The latest version of the design now fully supports this.
* If this is the first context being initialized, the native app will have the ability to get values of all computed properties (And change them).
* If it is a secondary context (CoreCLR is already loaded in the process), the native app can read all the properties which are coming from the secondary context (from its runtime config), as well as all the properties from the first context - and compare. There's no way to modify anything anymore (runtime doesn't support this), but the native app can decide if it's OK to use the secondary context or not (Given the property differences).